File: policysearchengine/policysearchengine-main/dao/es/es.py
from elasticsearch import Elasticsearch

import json
import requests
from datetime import datetime
from typing import List
import yaml
import warnings

# warnings.filterwarnings("ignore")

query_fmt = '''
	{
	  "query": {
	    "bool": {
	      "must": [
	        {
	          "bool": {
	            "should": [
	              {
	                "match": {
	                  "title": "%s"
	                }
	              },
	              {
	                "match": {
	                  "content": "%s"
	                }
	              }
	            ]
	          }
	        },
	        { "bool": {"must": [%s]} }
	        %s
	      ]
	    }
	  }
	}
	'''
ResultFmtDSL = '''
	{
	  %s
	  "sort": [
	    {
	      "date": {
	        "order": "asc"
	      }
	    }
	  ],
	  "from": %d,
	  "size": %d,
	  "highlight": %s
	}
	'''

# 构建高亮查询
highlight = '''
	{
	  "fields": {
	    "title": {},
	    "content": {}
	  },
	  "fragment_size": 50,
	  "pre_tags": ["<em style='color:red'>"],
	  "post_tags": ["</em>"]
	}
	'''

# 分数搜索DSL
fmtScoreQuery = '''
	{
	  "query": {
	    "function_score": {
	      %s
	      "functions": [
	        %s
	      ],
	      "score_mode": "multiply",
	      "boost_mode": "multiply"
	    }
	  },
	  "from": %d,
	  "size": %d,
	  "highlight": %s
	}'''

fmtScoreFilter = '''
	{
	  "filter": {
	    "bool": {
	      "should": [
	        { "match": { "title": "%s" }},
	        { "match": { "content": "%s" }}
	      ]
	    }
	  },
	  "weight": %f
	}'''

fmtMustDsl = '''
	{
	  "bool": {
	    "should": [
	      { "match": { "title": "%s" }},
	      { "match": { "content": "%s" }}
	    ],
	    "minimum_should_match": 1
	  }
	}'''

exactFmt = '''
	{
	  "bool": {
	    "should": [{
	      "match_phrase": {
	        "title": "%s"
	      }
	    }, {
	      "match_phrase": {
	        "content": "%s"
	      }
	    }]
	  }
	}'''


class ESDocument:
    def __init__(self, title: str, url: str, date: datetime, content: str,
                 department_id: int, sdp_ids: List[int], province_id: int):
        self.title = title
        self.url = url
        self.date = date.isoformat()  # 将日期转换为 ISO 格式字符串
        self.content = content
        self.department_id = department_id
        self.sdp_ids = sdp_ids
        self.province_id = province_id


# def init_es_client(config_path=None):
#     """初始化并返回Elasticsearch客户端"""
#     # 初始化配置解析器
#     config = configparser.ConfigParser()
#     # 读取配置文件
#     config.read(config_path)
#     # 从配置文件中获取Elasticsearch配置
#     es_host = config.get("es", "addr")
#     # es_user = config.get('elasticsearch', 'ES_USER')
#     # es_password = config.get('elasticsearch', 'ES_PASSWORD')
#
#     es = Elasticsearch(
#         hosts=[es_host],
#         # basic_auth=(es_user, es_password),
#         verify_certs=False,
#         ca_certs='conf/http_ca.crt'
#     )
#     return es
def init(config_path:str):
    with open(config_path, 'r',encoding='utf-8') as file:
        config = yaml.safe_load(file)  # 读取 YAML 文件并解析为字典

    es_address = config['es']['addr']  # 获取 Elasticsearch 地址
    # es_index = config['es']['index']  # 获取索引名称

    es = Elasticsearch([es_address])  # 创建 Elasticsearch 客户端

    if not es.ping():
        print(f"ES初始化错误，无法连接到 {es_address}")
        return 0

    print(f"ES客户端已初始化，连接到 {es_address}")
    return es  # 返回 ES 客户端


def create_index(es, index_name="test-index"):
    """创建索引，如果索引已存在则忽略"""
    if not es.indices.exists(index=index_name):
        es.indices.create(index=index_name)


def define_mapping(es, index_name="test-index"):
    """为索引定义映射"""
    mapping = {
        "mappings": {
            "properties": {
                "name": {"type": "text"},
                "age": {"type": "integer"},
                "email": {"type": "keyword"}
            }
        }
    }
    es.indices.create(index=index_name, body=mapping, ignore=400)  # ignore=400忽略索引已存在错误


def insert_document(es, index_name="test-index", doc_id=None, document=None):
    """插入文档到指定索引"""
    es.index(index=index_name, id=doc_id, document=document)


def update_document(es, index_name="test-index", doc_id=None, updated_doc=None):
    """更新指定ID的文档"""
    es.update(index=index_name, id=doc_id, body={"doc": updated_doc})


def delete_document(es, index_name="test-index", doc_id=None):
    """删除指定ID的文档"""
    es.delete(index=index_name, id=doc_id)


def search_documents(es, index_name="test-index", query=None):
    """在指定索引中搜索文档"""
    return es.search(index=index_name, body=query)


def index_doc(date: datetime, department_id: int, province_id: int,
              title: str, url: str, content: str, sdp_ids: List[int], index: str):
    doc = ESDocument(title, url, date, content, department_id, sdp_ids, province_id)

    data = json.dumps(doc.__dict__)  # 将文档对象转换为 JSON 格式
    # index = 'your_index_name'  # 这里替换为你的目标索引名称
    es_url = f'http://localhost:9200/{index}/_doc'  # Elasticsearch 文档上传的 URL

    response = requests.post(es_url, data=data, headers={"Content-Type": "application/json"})  # 上传数据

    if response.status_code != 201:
        print(f"ES上传数据失败，err: {response.text}")
        return

    print(response.json())  # 打印索引响应的内容

File: policy-search_engine/dao/es/es.go
package es

import (
	"PolicySearchEngine/config"
	"PolicySearchEngine/model"
	"bytes"
	"encoding/json"
	"fmt"
	"github.com/elastic/go-elasticsearch/v8"
	"strings"
	"time"
)

type SearchInput struct {
	Text       string //搜索内容？
	UseScore   bool
	ScoreField map[string]float64 //权重分数？
}

var (
	es    *elasticsearch.Client //es 将被用来持有一个 Elasticsearch 客户端的实例
	index string                //用来存储一个索引的名称。在 Elasticsearch 中，索引是存储相关数据的地方，类似于关系数据库中的表。
)

const (
	queryFmt = `
  "query": {
    "bool": {
      "must": [
        {
          "bool": {
            "should": [
              {
                "match": {
                  "title": "%s"
                }
              },
              {
                "match": {
                  "content": "%s"
                }
              }
            ]
          }
        },
		{ "bool": {"must": [%s]} }
		 %s
      ]
    }
  },
`
	ResultFmtDSL = `
	{
  %s
  "sort": [
    {
      "date": {
        "order": "asc"
      }
    }
  ],
  "from": %d,
  "size": %d,
  "highlight": %s
}
`

	// 构建高亮查询
	highlight = `
	{
		"fields": {
			"title": {},
			"content": {}
		},
		"fragment_size": 50,
		"pre_tags": ["<em style='color:red'>"],
		"post_tags": ["</em>"]
	}`

	// 分数搜索DSL
	fmtScoreQuery = `
	{
	  "query": {
	    "function_score": {
	      %s
	      "functions": [
			%s
	      ],
	      "score_mode": "multiply",
	      "boost_mode": "multiply"
	    }
	  },
		"from": %d,
		"size": %d,
		"highlight": %s
	}`

	fmtScoreFilter = `
	{
	  "filter": {
	    "bool": {
	      "should": [
	        { "match": { "title": "%s" }},
	        { "match": { "content": "%s" }}
	      ]
	    }
	  },
	  "weight": %f
	}`

	fmtMustDsl = `
            {
              "bool": {
                "should": [
                  { "match": { "title": "%s" }},
                  { "match": { "content": "%s" }}
                ],
                "minimum_should_match": 1
              }
            }`

	exactFmt = `{
					"bool": {
						"should": [{
							"match_phrase": {
								"title": "%s"
							}
						}, {
							"match_phrase": {
								"content": "%s"
							}
						}]
					}
				}`
)

func Init() {
	var err error
	cfg := elasticsearch.Config{
		Addresses: []string{
			config.V.GetString("es.addr"), //Addresses: []string{"http://localhost:9200"}
		},
	}
	es, err = elasticsearch.NewClient(cfg)
	if err != nil {
		fmt.Printf("ES初始化错误，err:%+v\n", err.Error())
	}
	index = config.V.GetString("es.index")
} //貌似是创建客户端

func IndexDoc(date time.Time, departmentID, provinceID uint, title, url, content string, sdpIDs []uint) {
	doc := model.ESDocument{
		Title:             title,
		Url:               url,
		Date:              date,
		Content:           content,
		DepartmentID:      departmentID,
		SmallDepartmentID: sdpIDs,
		ProvinceID:        provinceID,
	} //这是一个结构体，用于表示Elasticsearch文档。
	data, _ := json.Marshal(doc)                       //将doc结构体序列化为JSON格式，结果存储在data变量中
	idx, err := es.Index(index, bytes.NewReader(data)) //使用 Elasticsearch 客户端的 Index 方法将 JSON 数据上传到指定的索引中。index 是目标索引的名称，bytes.NewReader(data) 将字节数组转换为可读的 io.Reader 类型。
	if err != nil {
		fmt.Printf("ES上传数据失败，err:%+v\n", err)
		return
	}
	fmt.Println(idx.String()) //打印索引响应的内容。
} //上传es数据

func MatchAllDoc() {
	query := `{ "query": { "match_all": {} } }`
	search, err := es.Search(
		es.Search.WithIndex(index),
		es.Search.WithBody(strings.NewReader(query)),
	)
	if err != nil {
		return
	}
	fmt.Println(search.String())
}

// SearchDocWithSmallDepartmentID 根据小部门筛
func SearchDocWithSmallDepartmentID(searchQuery SearchInput, exact []string, smallDepartmentID, provinceID, from, size int) *model.ESResp {
	var query string
	searchFmt := queryFmtPrint(searchQuery.Text, exact, smallDepartmentID, provinceID)
	if !searchQuery.UseScore { //如果UseScore为false，则使用fmt.Sprintf函数和ResultFmtDSL格式化字符串来构建查询
		query = fmt.Sprintf(ResultFmtDSL,
			searchFmt,
			from-1,
			size,
			highlight)
	} else {
		query = fmt.Sprintf(fmtScoreQuery,
			searchFmt,
			fmtScoreFilters(searchQuery.ScoreField),
			from-1,
			size,
			highlight)
	}

	fmt.Println(query)

	searchResult, err := es.Search(
		es.Search.WithIndex(index),                   //指定要查询的Elasticsearch索引的名称。
		es.Search.WithBody(strings.NewReader(query)), //指定查询的请求体。
		es.Search.WithTrackTotalHits(true),           //启用跟踪总命中数的选项。
	)
	if err != nil {
		fmt.Println("Error executing search:", err)
		return nil
	}

	// 解析 searchResult 中的 JSON 数据
	var responseData model.ESResp
	_ = json.NewDecoder(searchResult.Body).Decode(&responseData) //使用json.NewDecoder解码器来解析searchResult.Body中的JSON数据，并将其存储在responseData变量中。忽略任何解码错误。

	return &responseData
}

func queryFmtPrint(text string, exact []string, smallDepartmentID, provinceID int) string {
	query := ","
	if smallDepartmentID == 0 && provinceID == 0 {
		query = ""
	} else if smallDepartmentID != 0 && provinceID != 0 {
		query += fmt.Sprintf(`{ "match": { "small_department_id": %d }},`, smallDepartmentID)
		query += fmt.Sprintf(`{ "match": { "province_id": %d }}`, provinceID)
	} else if smallDepartmentID != 0 && provinceID == 0 {
		query += fmt.Sprintf(`{ "match": { "small_department_id": %d }}`, smallDepartmentID)
	} else if smallDepartmentID == 0 && provinceID != 0 {
		query += fmt.Sprintf(`{ "match": { "province_id": %d }}`, provinceID)
	}

	var exactTrim []string
	for _, str := range exact {
		if str != "" {
			exactTrim = append(exactTrim, str)
		} //去除空字符串，得到 exactTrim 列表。
	}

	exactQuery := ""
	for i := 0; i < len(exactTrim); i++ {
		exactQuery += fmt.Sprintf(exactFmt, exactTrim[i])
		if i < len(exactTrim)-1 {
			exactQuery += ","
		}
	} //遍历 exactTrim 列表，构建精确匹配的查询字符串，并在每个条件之间添加逗号分隔。

	return fmt.Sprintf(queryFmt, text, text, exactQuery, query)
} //构建一个用于 Elasticsearch 查询的格式化字符串。通过不同的条件判断和字符串拼接，生成最终的查询字符串。

func fmtScoreFilters(m map[string]float64) string {
	var filters []string
	for word, weight := range m {
		filter := fmt.Sprintf(fmtScoreFilter, word, word, weight)
		filters = append(filters, filter)
	}
	return strings.Join(filters, ", ")
}
